job_controller/database/db_updater.py

```diff
@@ -0,0 +1,276 @@
+    """
+
+    def __init__(self, ip: str, username: str, password: str):
+        connection_string = f"postgresql+psycopg2://{username}:{password}@{ip}:5432/interactive-reduction"
+        engine = create_engine(connection_string, poolclass=QueuePool, pool_size=20)
+        self.session_maker_func = sessionmaker(bind=engine)
+        self.runs_table = Run()
+        self.reductions_table = Reduction()
+        self.runs_reductions_table = RunReduction()
+        self.script_table = Script()
+
+    # pylint: disable=too-many-arguments
+    def add_detected_run(
+        self,
+        filename: str,
+        title: str,
+        users: str,
+        experiment_number: str,
+        run_start: str,
+        run_end: str,
+        good_frames: str,
+        raw_frames: str,
+        reduction_inputs: Dict[str, Any],
+    ) -> int:
+        """
+        This function submits data to the database from what is initially available on detected-runs kafka topic
+        :param filename: the filename of the run that needs to be reduced
+        :param title: The title of the run file
+        :param users: The users entered into the run file
+        :param experiment_number: The RB number of the run entered by users
+        :param run_start: The time at which the run started, created using the standard python time format.
+        :param run_end: The time at which the run ended, created using the standard python time format.
+        :param good_frames: The number of frames that are considered "good" in the file
+        :param raw_frames: The number of frames that are in the file
+        :param reduction_inputs: The inputs to be used by the reduction
+        :return: The id of the reduction row entry
+        """
+        logger.info(
+            "Submitting detected-run to the database: {filename: %s, title: %s, users: %s, "
+            "experiment_number: %s, run_start: %s, run_end: %s, good_frames: %s, raw_frames: %s, "
+            "reduction_inputs: %s}",
+            filename,
+            title,
+            users,
+            experiment_number,
+            run_start,
+            run_end,
+            good_frames,
+            raw_frames,
+            reduction_inputs,
+        )
+        session = self.session_maker_func()
+        run = Run(
+            filename=filename,
+            title=title,
+            users=users,
+            experiment_number=experiment_number,
+            run_start=run_start,
+            run_end=run_end,
+            good_frames=good_frames,
+            raw_frames=raw_frames,
+        )
+        reduction = Reduction(
+            reduction_start=None,
+            reduction_end=None,
+            reduction_state=None,
+            reduction_inputs=reduction_inputs,
+            script=None,
+            reduction_outputs=None,
+        )
+        session.add(run)
+        session.add(reduction)
+        session.commit()
```

Comment:
Is there a reason for committing here? Would removing this not make this function atomic in the DB sense, currently it could potentially leave the DB part mutated.

Reply:
It needs to be committed to generate the ID, which is then used for the RunReduction row, I assumed it might sort this for us without committing beforehand and it all gets sorted by the ORM, but when I did that, it didn't function.